lib/awful/placement.lua

```diff
@@ -499,6 +498,37 @@ wibox_update_strut = function(d, position, args)
     d:struts(struts)
 end
 
+--- Return true if a **client** is on its screen edge.
+-- The `mouse.snap` code already handle the case it is moved with the mouse,
+-- but it stills need to be docked when placed programatically.
+-- @treturn boolean If the client touches the edge
+local function is_on_edge(d)
+    local geo  = geometry_common(d, {})
+
+    -- No need to call `get_parent_geometry`, this is always used on clients.
+    local sgeo = d.screen.geometry
+
+    return (geo.x == sgeo.x and "left")
+        or (geo.y == sgeo.y and "top")
+        or ((geo.x+geo.width ) == (sgeo.x+sgeo.width ) and "right")
+        or ((geo.y+geo.height) == (sgeo.y+sgeo.height) and "bottom") or nil
+end
+
+local function detect_and_update_struts(c)
+    if c.dockable then
+        local edge = is_on_edge(c)
+        if edge then
+            wibox_update_strut(c, edge, {})
+        else
+            c:struts { left = 0, right = 0, bottom = 0, top = 0 }
+        end
+    end
+end
+
+-- Listen to dockable clients geometry change and update the workarea.
+capi.client.connect_signal("property::geometry", detect_and_update_struts)
+capi.client.connect_signal("property::dockable", detect_and_update_struts)
```

Comment:
So, any client which is moved to the edge of the screen (for example by a layout) gets docked there? That doesn't seem right.

Also, I never really understood the magic here, but I thought the idea was to first compute the new geometry and only apply it in the end. So, shouldn't the struts only be applied in the end, too? That way they would not interfere with `no_offscreen`, because no struts are applied yet.
I don't really understand why #1449 happens then, but from your rules it is wrong, since it uses a callback to call into `awful.placement` even though it should use the `placement` key. No?